mapper-processor/src/test/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScannerTest.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.mapper.processor.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.datastax.oss.driver.api.mapper.annotations.HierarchyScanStrategy;
+import com.datastax.oss.driver.shaded.guava.common.collect.Lists;
+import java.util.Arrays;
+import java.util.List;
+import javax.lang.model.element.Name;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class HierarchyScannerTest {
+
+  Types types = Mockito.mock(Types.class);
+
+  @Test
+  public void should_build_proper_hierarchy_with_default_strategy() {
+    /*
+     * given the following hierarchy
+     *
+     *                  a
+     *                 /
+     *                z
+     *               /
+     *         a    y
+     *        /    /
+     *        b   x w
+     *         \ / /
+     *          c
+     */
+    MockInterface a = i("a");
+    MockInterface z = i("z", a);
+    MockInterface y = i("y", z);
+    MockInterface x = i("x", y);
+    MockInterface w = i("w");
+    MockClass b = c("b", null, a);
+    MockClass c = c("c", b, x, w);
+
+    // when no HierarchyScanStrategy is defined, then the default behavior should be used
+    // which is to scan the entire tree and return it in the correct (bottom up) order.
+    assertThat(HierarchyScanner.resolveTypeHierarchy(c.classElement, types).toString())
+        .isEqualTo("[c, b, x, w, a, y, z]");
```

Comment:
is the ordering of `a, y, z` is not important? Shouldn't it be `y,z,a`?

Reply:
a occurs before y and z in this context because b extends a.

If you look at the comment above with the graph the order of elements traversed should match the graph reading from bottom up, and left to right within the same depth of the graph, so for this case the order is encountered as

c
b x w
a y
z

since a has already been encountered, it's not added again.